provd/internal/services/user/user.go

```diff
@@ -0,0 +1,293 @@
+		userFactory: opts.userFactory,
+	}
+}
+
+func hashPassword(password string) (string, error) {
+	// Generate a 16-byte salt
+	salt := make([]byte, 16)
+	if _, err := rand.Read(salt); err != nil {
+		return "", err
+	}
+
+	// Hash the password with the salt
+	hasher := sha512.New()
+	hasher.Write(salt)
+	hasher.Write([]byte(password))
+	hashedBytes := hasher.Sum(nil)
+	hashedPassword := base64.RawStdEncoding.EncodeToString(hashedBytes)
+
+	// Format the hash in the SHA-512 crypt format
+	saltStr := base64.RawStdEncoding.EncodeToString(salt)
+	return fmt.Sprintf("$6$%s$%s", saltStr, hashedPassword), nil
+}
+
+func (s *Service) CreateUser(ctx context.Context, req *proto.CreateUserRequest) (*proto.Empty, error) {
+
+	// Validate requtest
+	if req == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "received a nil request")
+	}
+	user := req.GetUser()
+	if user == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "received a nil user")
+	}
+	username := user.GetUsername()
+	if username == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "recieved an empty username")
+	}
+	realName := user.GetRealName()
+	if realName == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "recieved an empty realName")
+	}
+	isAdmin := req.GetIsAdmin()
+	var accountType int32
+	if isAdmin {
+		accountType = 1
+	} else {
+		accountType = 0
+	}
+	password := user.GetPassword() // allow empty password?
+	autologin := user.GetAutoLogin()
+	hostname := user.GetHostname()
+	if hostname == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "recieved an empty hostname")
+	}
+	// Create the user
+	var userObjectPath dbus.ObjectPath
+	call := s.accounts.Call("org.freedesktop.Accounts.CreateUser", 0, username, realName, accountType)
+
+	err := call.Store(&userObjectPath)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to create user: %s", err)
+	}
+
+	hashed, err := hashPassword(password)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to generate hashed password: %s", err)
+	}
+
+	// Set the password for the user
+	userObject := s.userFactory.GetUserObject(userObjectPath)
+	fmt.Printf(hashed)
+	err = userObject.Call("org.freedesktop.Accounts.User.SetPassword", 0, hashed, "").Err
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to set password: %s", err)
+	}
+
+	// Set autologin for the user
+	err = userObject.Call("org.freedesktop.Accounts.User.SetAutomaticLogin", 0, autologin).Err
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to set autologin: %s", err)
+	}
+
+	// Set the hostname
+	err = s.hostname.Call("org.freedesktop.hostname1.SetStaticHostname", 0, hostname, false).Err
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to set hostname: %s", err)
+	}
+
+	return &proto.Empty{}, nil
+}
+
+func (s *Service) ValidateUsername(ctx context.Context, req *proto.ValidateUsernameRequest) (*proto.ValidateUsernameResponse, error) {
```

Comment:
This method only checks if the given username already exists, but there are a few more cases we need to consider, see [implementation in subiquity](https://github.com/canonical/subiquity/blob/e3a0f6d215fdd19deaa854bdec0298b4a5b57ee3/subiquity/server/controllers/identity.py#L122-L135).

I'm wondering what's the best way to return the validation result - a boolean plus a specific error in case of failure, or an enum like in subiquity?

Reply:
I've updated the impl to check those additional restrictions :+1: 

Now that it accounts for more failure cases, using an emun to represent the outcome makes sense to me :slightly_smiling_face: we are going to use the same pattern for ad validation aswell.